runtime-modules/project-token/src/types.rs

```diff
@@ -0,0 +1,414 @@
+impl<Balance: Zero + Copy + PartialOrd + Saturating, ReserveBalance>
+    AccountData<Balance, ReserveBalance>
+{
+    /// Ctor
+    pub fn new_with_liquidity_and_bond(
+        init_liquidity: Balance,
+        bloat_bond: ReserveBalance,
+    ) -> Self {
+        AccountData::<_, _> {
+            free_balance: init_liquidity,
+            stacked_balance: Balance::zero(),
+            bloat_bond,
+        }
+    }
+
+    /// Check wheather account is empty
+    pub(crate) fn is_empty(&self) -> bool {
+        self.total_balance().is_zero()
+    }
+
+    /// Increase liquidity for an account
+    pub(crate) fn increase_liquidity_by(&mut self, amount: Balance) {
+        self.free_balance = self.free_balance.saturating_add(amount);
+    }
+
+    /// Decrease liquidity for an account
+    pub(crate) fn decrease_liquidity_by(&mut self, amount: Balance) {
+        self.free_balance = self.free_balance.saturating_sub(amount);
+    }
+
+    /// Dry run for `self.decrease_liquidity_by(amount)`
+    pub(crate) fn ensure_can_decrease_liquidity_by<T: crate::Trait>(
+        &self,
+        amount: Balance,
+    ) -> DispatchResult {
+        ensure!(
+            self.free_balance >= amount,
+            crate::Error::<T>::InsufficientFreeBalance,
+        );
+        Ok(())
+    }
+
+    pub(crate) fn total_balance(&self) -> Balance {
+        self.free_balance.saturating_add(self.stacked_balance)
+    }
+}
+/// Token Data implementation
+impl<
+        Balance: Zero + Copy + Saturating + Debug + From<u64> + UniqueSaturatedInto<u64> + Unsigned,
+        Hash,
+        BlockNumber: Copy + Saturating + PartialOrd + AtLeast32BitUnsigned,
+    > TokenData<Balance, Hash, BlockNumber>
+{
+    // increment account number
+    pub(crate) fn increment_accounts_number(&mut self) {
+        self.accounts_number = self.accounts_number.saturating_add(1u64);
+    }
+
+    // decrement account number
+    pub(crate) fn decrement_accounts_number(&mut self) {
+        self.accounts_number = self.accounts_number.saturating_sub(1u64);
+    }
+
+    // increase total supply
+    pub(crate) fn increase_supply_by(&mut self, amount: Balance) {
+        self.supply = self.supply.saturating_add(amount);
+    }
+
+    // decrease total supply
+    pub(crate) fn decrease_supply_by(&mut self, amount: Balance) {
+        self.supply = self.supply.saturating_sub(amount);
+    }
+
+    pub fn set_unclaimed_tally_patronage_at_block(&mut self, amount: Balance, block: BlockNumber) {
+        self.patronage_info.last_unclaimed_patronage_tally_block = block;
+        self.patronage_info.unclaimed_patronage_tally_amount = amount;
+    }
+
+    /// Computes: period * rate * supply + tally
+    pub(crate) fn unclaimed_patronage_at_block(&self, block: BlockNumber) -> Balance {
+        let blocks = block.saturating_sub(self.patronage_info.last_unclaimed_patronage_tally_block);
+        let unclaimed_patronage_percent = self.patronage_info.rate.for_period(blocks);
```

Comment:
Just note that if unclaimed patronage exceeds 100% of the supply, `unclaimed_patronage_percent` will still be  `100%` (since `Perquintill` cannot be more than 100%).

I don't think this is an issue, because in reality this will probably never happen, but perhaps worth introducing a test case for that (just to make sure it doesn't panic or cause some unexpected state)